File: developer_packages/rom2109/rom2109_switch/scripts/switch_mode.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import os
import subprocess
import time

from ament_index_python.packages import get_package_share_directory

class SwitchMode(Node):
    
    def __init__(self):
        super().__init__('switch_mode')
        self.current_process = None
        self.current_mode = None  
        rom_robot_name = os.environ.get('ROM_ROBOT_MODEL', 'rom2109')

        nav2_pkg = get_package_share_directory(f'{rom_robot_name}_nav2')

        self.navigation_launch_file   = os.path.join( nav2_pkg, 'launch', 'navigation.launch.py')
        self.cartographer_launch_file = os.path.join( nav2_pkg, 'launch', 'mapping.launch.py')
        self.remapping_launch_file    = os.path.join( nav2_pkg, 'launch', 'mapping.launch.py')

        # Create a subscriber
        self.subscription = self.create_subscription(
            String,
            'switch_mode',
            self.mode_callback,
            10
        )

    def mode_callback(self, msg):
        new_mode = msg.data

        if new_mode == self.current_mode:
            self.get_logger().info(f"Mode is already {new_mode}. No change needed.")
            return

        if self.current_process is not None:
            self.get_logger().info(f"Shutting down current process for mode: {self.current_mode}")
            self.current_process.terminate()  
            self.current_process.wait()  
            self.current_process = None  
            self.get_logger().info(f"Process for {self.current_mode} terminated.")

        if new_mode == 'navi':
            self.get_logger().info('Switching to navigation mode...')
            self.current_process = subprocess.Popen(['ros2', 'launch', 'rom2109_nav2', 'navigation.launch.py'])
            self.current_mode = new_mode

        elif new_mode == 'mapping':
            self.get_logger().info('Switching to mapping mode...')
            self.current_process = subprocess.Popen(['ros2', 'launch', 'rom2109_nav2', 'mapping.launch.py'])
            self.current_mode = new_mode

        elif new_mode == 'remapping':
            self.get_logger().info('Switching to remapping mode...')
            self.current_process = subprocess.Popen(['ros2', 'launch', 'rom2109_nav2', 'remapping.launch.py'])
            self.current_mode = new_mode

        self.get_logger().info(f"New mode '{new_mode}' process started.")

def main(args=None):
    rclpy.init(args=args)
    node = SwitchMode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("Shutting down...")

    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: developer_packages/rom2109/rom2109_switch/src/switch_mode.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <cstdlib>
#include <string>
#include <memory>
#include <vector>
#include <signal.h>
#include <sys/wait.h>
#include <unistd.h>

class SwitchMode : public rclcpp::Node {
public:
    SwitchMode() : Node("switch_mode"), current_process_pid(-1), current_mode("") {
        // Get robot model from environment variable or use default
        const char *env_p = std::getenv("ROM_ROBOT_MODEL");
        robot_model = (env_p != nullptr) ? std::string(env_p) : "rom2109";

        // Initialize paths to launch files
        nav2_launch_file = robot_model + "_nav2/navigation.launch.py";
        mapping_launch_file = robot_model + "_nav2/mapping.launch.py";
        remapping_launch_file = robot_model + "_nav2/remapping.launch.py";

        // Subscription to the "switch_mode" topic
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "switch_mode",
            10,
            std::bind(&SwitchMode::modeCallback, this, std::placeholders::_1)
        );
    }

    ~SwitchMode() {
        // Ensure the process is cleaned up on node destruction
        terminateCurrentProcess();
    }

private:
    void modeCallback(const std_msgs::msg::String::SharedPtr msg) {
        std::string new_mode = msg->data;

        if (new_mode == current_mode) {
            RCLCPP_INFO(this->get_logger(), "Mode is already '%s'. No change needed.", new_mode.c_str());
            return;
        }

        // Terminate the current process if any
        terminateCurrentProcess();

        // Launch the new process based on the requested mode
        if (new_mode == "navi") {
            RCLCPP_INFO(this->get_logger(), "Switching to navigation mode...");
            launchProcess(nav2_launch_file);
        } else if (new_mode == "mapping") {
            RCLCPP_INFO(this->get_logger(), "Switching to mapping mode...");
            launchProcess(mapping_launch_file);
        } else if (new_mode == "remapping") {
            RCLCPP_INFO(this->get_logger(), "Switching to remapping mode...");
            launchProcess(remapping_launch_file);
        } else {
            RCLCPP_WARN(this->get_logger(), "Invalid mode: '%s'. No action taken.", new_mode.c_str());
        }

        current_mode = new_mode;
    }

    void launchProcess(const std::string &launch_file) {
        // Fork and execute the launch file in the child process
        pid_t pid = fork();
        if (pid == 0) {
            // Child process
            execlp("ros2", "ros2", "launch", launch_file.c_str(), nullptr);
            // If execlp fails
            perror("execlp");
            std::exit(EXIT_FAILURE);
        } else if (pid > 0) {
            // Parent process
            current_process_pid = pid;
            RCLCPP_INFO(this->get_logger(), "Launched process with PID: %d", pid);
        } else {
            // Fork failed
            RCLCPP_ERROR(this->get_logger(), "Failed to fork a new process.");
        }
    }

    void terminateCurrentProcess() {
        if (current_process_pid > 0) {
            RCLCPP_INFO(this->get_logger(), "Terminating process with PID: %d", current_process_pid);
            kill(current_process_pid, SIGTERM);
            waitpid(current_process_pid, nullptr, 0);  // Wait for the process to terminate
            current_process_pid = -1;
            RCLCPP_INFO(this->get_logger(), "Process terminated.");
        }
    }

    // Member variables
    std::string robot_model;
    std::string nav2_launch_file;
    std::string mapping_launch_file;
    std::string remapping_launch_file;
    std::string current_mode;
    pid_t current_process_pid;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);

    auto node = std::make_shared<SwitchMode>();
    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}
